File: Algorithms/23. Merge k Sorted Lists - Method 1.py
# You are given an array of k linked-lists lists, each linked-list is sorted in ascending order.

# Merge all the linked-lists into one sorted linked-list and return it.


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        nodes = []
        heapq.heapify(nodes)
        for i in range(len(lists)):
            if lists[i]:
                heapq.heappush(nodes, (lists[i].val, i))
                lists[i] = lists[i].next
        
        output = ListNode(0)
        cur = output

        while nodes:
            node, index = heapq.heappop(nodes)
            cur.next = ListNode(node)
            cur = cur.next
            if lists[index]:
                heapq.heappush(nodes, (lists[index].val, index))
                lists[index] = lists[index].next
        
        return output.next

File: Algorithms/23. Merge k Sorted Lists - Method 1.java
// You are given an array of k linked-lists lists, each linked-list is sorted in ascending order.

// Merge all the linked-lists into one sorted linked-list and return it.


/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        Queue<ListNode> nodes = new PriorityQueue<>((n1, n2) -> {return n1.val - n2.val;});
        for (int i = 0; i < lists.length; i++) {
            if (lists[i] != null) {
                nodes.add(lists[i]);
                lists[i] = lists[i].next;
            }
        }

        ListNode output = new ListNode(0);
        ListNode cur = output;

        while (!nodes.isEmpty()) {
            ListNode node = nodes.poll();
            cur.next = node;
            cur = cur.next;
            node = node.next;

            if (node != null) {
                nodes.add(node);
            }
        }

        return output.next;
    }
}
